Source/DafnyDriver/DafnyDriver.cs

```diff
@@ -597,8 +597,9 @@ class DafnyConsolePrinter : ConsolePrinter {
 
     #region Compilation
 
-    static string WriteDafnyProgramToFiles(Compiler compiler, string dafnyProgramName, bool targetProgramHasErrors,
-      string targetProgramText, string/*?*/ callToMain, Dictionary<string, string> otherFiles, TextWriter outputWriter) {
+    private record Locations(string targetDir, string targetFilename, string relativeTarget, string targetBaseDir);
```

Comment:
`Locations` seems like too generic of a name to me. Given that `target` is part of all the field names, it might be nicer to put it in the name of the record. 

Consider naming the record `TargetPaths` with the field names `directory` (was `targetDir`), `filename` (was `targetFilename`), `relativeFilename` (was `relativeTarget`), `relativeDirectory` (was `targetBaseDir`).

I find the (existing) name `targetBaseDir` confusing because it's not a prefix of the target directory, which I would expect given the term `base`.

Also consider only having the fields `filename` and `relativeFilename` since the directories can be quickly computed from the filename using `Path.GetDirectoryName`. 

My personal preference would even be to have `baseDirectory` and `filename` be the only fields, with `baseDirectory` defined so that `relativeFilename == Path.GetRelativePath(baseDirectory, filename)`. The reason I like this is because it has the minimum amount of path duplication in the fields of `Location`, but I'm not sure if it has practical advantages.